### PR TITLE
in progress flag for deploy

### PR DESCRIPTION
**Pull request by benrudolph**
@gcapalbo ran threw this pretty quickly, but doesn't quite work as i expected. django-redis seems to be prefixing its keys with `:1:`. and then when i try to set a key via `redis-cli` like `:1:deploy_in_progress` django-redis throws an error when trying to read it if it hasn't originally set it:

```
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/cchq/www/staging/releases/2015-11-03_16.53/python_env/local/lib/python2.7/site-packages/django_redis/cache.py", line 25, in _decorator
    return method(self, *args, **kwargs)
  File "/home/cchq/www/staging/releases/2015-11-03_16.53/python_env/local/lib/python2.7/site-packages/django_redis/cache.py", line 73, in get
    client=client)
  File "/home/cchq/www/staging/releases/2015-11-03_16.53/python_env/local/lib/python2.7/site-packages/django_redis/client/default.py", line 205, in get
    return self.decode(value)
  File "/home/cchq/www/staging/releases/2015-11-03_16.53/python_env/local/lib/python2.7/site-packages/django_redis/client/default.py", line 303, in decode
    value = self._serializer.loads(value)
  File "/home/cchq/www/staging/releases/2015-11-03_16.53/python_env/local/lib/python2.7/site-packages/django_redis/serializers/pickle.py", line 43, in loads
    return pickle.loads(force_bytes(value))
UnpicklingError: could not find MARK
```

nothing came up via google and was wondering if you had any insight.

cc: @proteusvacuum

### CODE REVIEW

**Comment by gcapalbo** [comment]
I've seen some weird behavior in the past where setting a value from the django cache api and setting a value with the raw client from django-redis stores it to two different places - sounds like this might be related.  Is there anyway to set the value from the normal django cache api? (management command maybe, but that's not great)

File: corehq/apps/hqwebapp/views.py [hunk 1]
```diff
@@ -83,6 +86,11 @@ def couch_check():
         return isinstance(results, list), None
 
 
+def is_deploy_in_progress():
+    cache = get_redis_default_cache()
```
Comment (gcapalbo) [inline corehq/apps/hqwebapp/views.py]: I think we'd also want to be explicit about this and only set / get this value from the redis cache (not any other default cache that might be defined), unless we can guarantee the same set and expire functionality when setting the flag for each cache that might be used.

File: fab/fabfile.py [hunk 2]
```diff
@@ -483,6 +486,21 @@ def record_successful_deploy(url):
         })
 
 
+@task
+@roles(ROLES_DB_ONLY)
+def set_in_progress_flag():
+    sudo('redis-cli SET {} true'.format(IN_PROGRESS_FLAG))
+
+    # Set deploy flag to expire in 30 minutes
+    sudo('redis-cli EXPIRE {} {}'.format(IN_PROGRESS_FLAG, 60 * 30))
```
Comment (gcapalbo) [inline fab/fabfile.py]: I think 30 minutes is too much (having celery be down for 30 minutes without knowing would be bad).  Can we set the flag right before we stop all celery services and set the timeout to 5-10 minutes?

Comment (benrudolph) [inline fab/fabfile.py]: sure i think we might want to change the flag name then cause it won't necessarily be true

File: corehq/apps/hqwebapp/views.py (hunk 1 above)
Comment (benrudolph) [inline corehq/apps/hqwebapp/views.py]: this just does the getting so we don't have to worry about the expires/setting here. i was leveraging the existing `CACHES` variable. direct connection makes sense to me, can do. we should probably surround in try catch as well and return false on error

**Comment by benrudolph** [comment]
dang ok, management command can work

File: corehq/apps/hqwebapp/views.py (hunk 1 above)
Comment (benrudolph) [inline corehq/apps/hqwebapp/views.py]: definitely should do if we switch the `redis-cli` to a mgmt command

**Comment by czue** [comment]
nice this is clever

File: fab/fabfile.py [hunk 3]
```diff
@@ -68,6 +68,9 @@
     '*.db',
     )
 RELEASE_RECORD = 'RELEASES.txt'
+# Django redis seems to uses this to prefix their keys
```
Comment (proteusvacuum) [inline fab/fabfile.py]: nit / typo: use

File: corehq/apps/hqwebapp/views.py [hunk 4]
```diff
@@ -96,9 +104,14 @@ def celery_check():
         else:
             chk = (True, None)
     except IOError as e:
+        if is_deploy_in_progress():
+            return (True, 'Currently down, but deploy is in progress')
```
Comment (proteusvacuum) [inline corehq/apps/hqwebapp/views.py]: not a huge deal, but elsewhere in this method `chk` is set instead of immediately returning.

**Comment by proteusvacuum** [comment]
Not 100% sure I know what the purpose of this is? So that pingdom doesn't ping us during a deploy?

**Comment by benrudolph** [comment]
yep that's correct. pingdom kind of loses its umph if we cry wolf everyday during deploy

File: corehq/apps/hqwebapp/views.py (hunk 4 above)
Comment (benrudolph) [inline corehq/apps/hqwebapp/views.py]: yeah it was more to avoid having to add extra `else` clauses

**Comment by benrudolph** [comment]
@gcapalbo here's a less invasive way of doing the redis key. i think it's ok to use `get_default_redis_cache` since we use it in both places. also i took out the unset since it expires after 5 minutes and it takes a while to fire up a management command. let me know if you're OK with that

**Comment by gcapalbo** [comment]
@benrudolph looks good - i think it's fine to leave out the unset if the timeout is just 5 minutes

**Comment by gcapalbo** [comment]
this is good to merge now right (still has the don't merge label)?

**Comment by benrudolph** [comment]
g2g thanks

**Pull request merged**